src/box/lua/xlog.c

```diff
@@ -161,17 +278,16 @@ lbox_xlog_parser_iterate(struct lua_State *L)
 {
 	struct xlog_cursor *cur = lbox_checkcursor(L, 1, "xlog:pairs()");
 
-	struct xrow_header row;
 	int rc = 0;
 	/* skip all bad read requests */
 	while (true) {
-		rc = xlog_cursor_next_row(cur, &row);
-		if (rc == 0)
-			break;
-		if (rc < 0) {
-			struct error *e = diag_last_error(diag_get());
-			if (e->type != &type_XlogError)
-				luaT_error(L);
+		assert(xlog_cursor_is_open(cur));
+		if (cur->state == XLOG_CURSOR_TX) {
```

Comment:
I don't like that we intrude into cursor private members here.
What about introducing cursor raw API like this?
```diff
diff --git a/src/box/xlog.c b/src/box/xlog.c
index 08ce2c763..889ebaf41 100644
--- a/src/box/xlog.c
+++ b/src/box/xlog.c
@@ -1827,6 +1827,16 @@ xlog_tx_cursor_next_row(struct xlog_tx_cursor *tx_cursor,
 	return 0;
 }
 
+int
+xlog_tx_cursor_next_row_raw(struct xlog_tx_cursor *tx_cursor,
+			    struct ibuf **buf)
+{
+	if (ibuf_used(&tx_cursor->rows) == 0)
+		return 1;
+	*buf = &tx_cursor->rows;
+	return 0;
+}
+
 int
 xlog_tx_cursor_destroy(struct xlog_tx_cursor *tx_cursor)
 {
@@ -1928,6 +1938,20 @@ xlog_cursor_next_row(struct xlog_cursor *cursor, struct xrow_header *xrow)
 	return rc;
 }
 
+int
+xlog_cursor_next_row_raw(struct xlog_cursor *cursor, struct ibuf **buf)
+{
+	assert(xlog_cursor_is_open(cursor));
+	if (cursor->state != XLOG_CURSOR_TX)
+		return 1;
+	int rc = xlog_tx_cursor_next_row_raw(&cursor->tx_cursor, buf);
+	if (rc != 0) {
+		cursor->state = XLOG_CURSOR_ACTIVE;
+		xlog_tx_cursor_destroy(&cursor->tx_cursor);
+	}
+	return rc;
+}
+
 int
 xlog_cursor_next(struct xlog_cursor *cursor,
 		 struct xrow_header *xrow, bool force_recovery)
```
The agreement would be the same - you must not call `xlog_cursor_next_tx` util `xlog_cursor_next_row*` will return nonzero, so you would break reading of current TX by setting `buf.rpos = buf.wpos` on any error..

Reply:
Done.

Note that to avoid a memory leak in case the caller fails to parse an xrow returned by `xlog_cursor_next_row_raw`, we also have to move the call to `xlog_tx_cursor_destroy` from `xlog_tx_cursor_next_row` to `xlog_cursor_next_tx`.

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/xlog.c b/src/box/lua/xlog.c
index ccdf15be1e39..43a6880f927e 100644
--- a/src/box/lua/xlog.c
+++ b/src/box/lua/xlog.c
@@ -159,13 +159,9 @@ lbox_xlog_parse_body(struct lua_State *L, int type, const char **beg)
  * On success returns the number of values pushed (> 0). On EOF returns 0.
  */
 static int
-lbox_xlog_parse_row(struct lua_State *L, struct ibuf *rows)
+lbox_xlog_parse_row(struct lua_State *L, const char **pos, const char *end)
 {
 	int top = lua_gettop(L);
-	if (ibuf_used(rows) == 0)
-		goto eof;
-	const char **pos = (const char **)&rows->rpos;
-	const char *end = rows->wpos;
 	const char *tmp = *pos;
 	if (mp_check(&tmp, end) != 0 || mp_typeof(**pos) != MP_MAP)
 		goto bad_row;
@@ -269,7 +265,6 @@ dump:
 bad_row:
 	/* Silently assume EOF on bad row. */
 	lua_settop(L, top);
-eof:
 	return 0;
 }
 
@@ -281,13 +276,14 @@ lbox_xlog_parser_iterate(struct lua_State *L)
 	int rc = 0;
 	/* skip all bad read requests */
 	while (true) {
-		assert(xlog_cursor_is_open(cur));
-		if (cur->state == XLOG_CURSOR_TX) {
-			rc = lbox_xlog_parse_row(L, &cur->tx_cursor.rows);
+		const char **data;
+		const char *end;
+		rc = xlog_cursor_next_row_raw(cur, &data, &end);
+		assert(rc >= 0);
+		if (rc == 0) {
+			rc = lbox_xlog_parse_row(L, data, end);
 			if (rc > 0)
 				return rc;
-			cur->state = XLOG_CURSOR_ACTIVE;
-			xlog_tx_cursor_destroy(&cur->tx_cursor);
 		}
 		while ((rc = xlog_cursor_next_tx(cur)) < 0) {
 			struct error *e = diag_last_error(diag_get());
diff --git a/src/box/xlog.c b/src/box/xlog.c
index fc088d20cdde..6efd210db9b8 100644
--- a/src/box/xlog.c
+++ b/src/box/xlog.c
@@ -1868,6 +1868,17 @@ xlog_tx_cursor_next_row(struct xlog_tx_cursor *tx_cursor,
 	return 0;
 }
 
+int
+xlog_tx_cursor_next_row_raw(struct xlog_tx_cursor *tx_cursor,
+			    const char ***data, const char **end)
+{
+	if (ibuf_used(&tx_cursor->rows) == 0)
+		return 1;
+	*data = (const char **)&tx_cursor->rows.rpos;
+	*end = (const char *)tx_cursor->rows.wpos;
+	return 0;
+}
+
 int
 xlog_tx_cursor_destroy(struct xlog_tx_cursor *tx_cursor)
 {
@@ -1908,7 +1919,10 @@ xlog_cursor_next_tx(struct xlog_cursor *i)
 {
 	int rc;
 	assert(xlog_cursor_is_open(i));
-
+	if (i->state == XLOG_CURSOR_TX) {
+		i->state = XLOG_CURSOR_ACTIVE;
+		xlog_tx_cursor_destroy(&i->tx_cursor);
+	}
 	/* load at least magic to check eof */
 	rc = xlog_cursor_ensure(i, sizeof(log_magic_t));
 	if (rc < 0)
@@ -1961,12 +1975,17 @@ xlog_cursor_next_row(struct xlog_cursor *cursor, struct xrow_header *xrow)
 	assert(xlog_cursor_is_open(cursor));
 	if (cursor->state != XLOG_CURSOR_TX)
 		return 1;
-	int rc = xlog_tx_cursor_next_row(&cursor->tx_cursor, xrow);
-	if (rc != 0) {
-		cursor->state = XLOG_CURSOR_ACTIVE;
-		xlog_tx_cursor_destroy(&cursor->tx_cursor);
-	}
-	return rc;
+	return xlog_tx_cursor_next_row(&cursor->tx_cursor, xrow);
+}
+
+int
+xlog_cursor_next_row_raw(struct xlog_cursor *cursor,
+			 const char ***data, const char **end)
+{
+	assert(xlog_cursor_is_open(cursor));
+	if (cursor->state != XLOG_CURSOR_TX)
+		return 1;
+	return xlog_tx_cursor_next_row_raw(&cursor->tx_cursor, data, end);
 }
 
 int
diff --git a/src/box/xlog.h b/src/box/xlog.h
index 1cc2934611a9..fbfcbd838dd1 100644
--- a/src/box/xlog.h
+++ b/src/box/xlog.h
@@ -595,11 +595,30 @@ xlog_tx_cursor_destroy(struct xlog_tx_cursor *tx_cursor);
  * Fetch next xrow from xlog tx cursor
  *
  * @retval 0 for Ok
+ * @retval 1 if current tx is done
  * @retval -1 for error
  */
 int
 xlog_tx_cursor_next_row(struct xlog_tx_cursor *tx_cursor, struct xrow_header *xrow);
 
+/**
+ * Fetch next xrow from current xlog tx cursor.
+ *
+ * This function is similar to xlog_tx_cursor_next_row() except it doesn't
+ * parse the xrow nor does it advance the data pointer.
+ *
+ * @param cursor cursor
+ * @param[out] data pointer to the position in the internal buffer where
+ *                  the next xrow is stored
+ * @param[out] end end of the buffer
+ *
+ * @retval 0 for Ok
+ * @retval 1 if current tx is done
+ */
+int
+xlog_tx_cursor_next_row_raw(struct xlog_tx_cursor *cursor,
+			    const char ***data, const char **end);
+
 /**
  * Return current tx cursor position
  *
@@ -757,6 +776,24 @@ xlog_cursor_next_tx(struct xlog_cursor *cursor);
 int
 xlog_cursor_next_row(struct xlog_cursor *cursor, struct xrow_header *xrow);
 
+/**
+ * Fetch next xrow from current xlog tx.
+ *
+ * This function is similar to xlog_cursor_next_row() except it doesn't
+ * parse the xrow nor does it advance the data pointer.
+ *
+ * @param cursor cursor
+ * @param[out] data pointer to the position in the internal buffer where
+ *                  the next xrow is stored
+ * @param[out] end end of the buffer
+ *
+ * @retval 0 for Ok
+ * @retval 1 if current tx is done
+ */
+int
+xlog_cursor_next_row_raw(struct xlog_cursor *cursor,
+			 const char ***data, const char **end);
+
 /**
  * Fetch next row from cursor, ignores xlog tx boundary,
  * open a next one tx if current is done.
```

</details>